src/librustc/ty/mod.rs

```diff
@@ -2059,9 +2059,10 @@ impl ReprOptions {
     }
 
     /// Returns `true` if this `#[repr()]` should inhibit struct field reordering
-    /// optimizations, such as with repr(C) or repr(packed(1)).
+    /// optimizations, such as with repr(C), repr(packed(1)), or repr(<int>).
     pub fn inhibit_struct_field_reordering_opt(&self) -> bool {
-        !(self.flags & ReprFlags::IS_UNOPTIMISABLE).is_empty() || (self.pack == 1)
+        self.flags.intersects(ReprFlags::IS_UNOPTIMISABLE) || self.pack == 1 ||
+            self.int.is_some()
```

Comment:
cc @rkruppe This seems somewhat fragile (I mean the pre-existing code).
The check for `pack == 1` means things will still be reordered with `#[repr(packed(N))]`, is that intentional?
Also, it seems to me that `inhibit_enum_layout_opt` is a subset of this function and maybe we should have just one function that checks all conditions?

Reply:
The RFC does not specify any behaviour /wrt `repr(packed(N))` for enums and I’m not sure I recall us guaranteeing that `repr(packed)` and the reference [seems to agree](https://doc.rust-lang.org/nightly/reference/type-layout.html#the-packed-representation).

Reply:
In fact as per strict reading of the reference, `self.pack == 1` inhibiting reordering would be somewhat wrong, as it should cause "packing" for the original ordering, but since the "default" ordering is not specified, it seems to be fine either way...